# <p style="background-color: #f5df18; padding: 10px;"> Foundations of Astronomical Data Science | **Join** </p>



### <strong>Instructor: <span style="color: darkblue;">Devontae C. Baxter</span></strong>

Estimated completion time: 🕚 90 minutes



<div style="display: flex;">
    <div style="flex: 1; margin-right: 100px;">
        <h2>Questions</h2>
        <ul>
            <li>How do we use JOIN to combine information from multiple tables?</li>
            <li>How can we make a selection within a joined table?</li>
            <li>How should we save the result?</li>
        </ul>
    </div>
    <div style="flex: 1;">
        <h2>Learning Objectives</h2>
        <ul>
            <li>Write ADQL queries involving JOIN operations.</li>
            <li>Save data in CSV format</li>
        </ul>
    </div>
</div>

This imports previously imported functions:

In [1]:
from astroquery.gaia import Gaia
import pandas as pd

from episode_functions import *

The following code loads in the data (instructions for downloading data can be
found in the [setup instructions](../learners/setup.md)).  You may need to add a the path
to the filename variable below (e.g. `filename = 'student_download/backup-data/gd1_data.hdf'`)


In [2]:
filename = 'gd1_data.hdf'
point_series = pd.read_hdf(filename, 'point_series')
sky_point_list = point_series['sky_point_list']
pmra_min = point_series['pmra_min']
pmra_max = point_series['pmra_max']
pmdec_min = point_series['pmdec_min']
pmdec_max = point_series['pmdec_max']
point_series

sky_point_list    135.306, 8.39862, 126.51, 13.4449, 163.017, 54...
pmra_min                                                       -6.7
pmra_max                                                         -3
pmdec_min                                                    -14.31
pmdec_max                                                     -11.2
dtype: object

The next step in our analysis is to
select candidate stars based on photometry data.
The following figure from the Price-Whelan and Bonaca paper is a color-magnitude diagram for
the stars selected based on proper motion:

<div>
<img src="https://datacarpentry.org/astronomy-python/fig/gd1-3.png" width="550"/>
</div>

In orange is a [stellar
isochrone](https://en.wikipedia.org/wiki/Stellar_isochrone), showing
where we expect the stars in GD-1 to fall based on the metallicity and
age of their original globular cluster.

By selecting stars in the shaded area, we can further distinguish the
main sequence of GD-1 from younger background stars.

# Outline
---

1. We will reload the candidate stars we identified in the previous
  episode.

2. Then we will run a query on the Gaia server that uploads the table
  of
  candidates and uses a `JOIN` operation to select photometry data for
  the candidate stars.

3. We will write the results to a file for use in the next episode.

# Getting photometry data
---

The Gaia dataset contains some photometry data, including the variable
`bp_rp`, which contains BP-RP color (the difference in mean flux
between the BP and RP bands).
We use this variable to select stars with `bp_rp` between -0.75 and 2,
which excludes many class M dwarf stars.

But we can do better than that. Assuming GD-1 is a globular cluster, all of the stars formed at the same
time from the same material, so the stars' photometric properties should be consistent with a single
isochrone in a color magnitude diagram.
We can use photometric color and apparent magnitude to select stars with the age and metal richness we expect in GD-1.
However, the broad Gaia photometric bands (G, BP, RP) are not optimized for this task, instead we will use the more narrow photometric bands available from the Pan-STARRS survey to obtain the `g-i` color and apparent `g`\-band magnitude.

Conveniently, the Gaia server provides data from Pan-STARRS as a table
in the same database we have been using, so we can access it by making
ADQL queries.

## 🔔 A caveat about matching stars between catalogs
---

In general, choosing a star from the Gaia catalog and finding the
corresponding star in the Pan-STARRS catalog is not easy.  This kind
of cross matching is not always possible, because a star might appear
in one catalog and not the other.  And even when both stars are
present, there might not be a clear one-to-one relationship between
stars in the two catalogs. Additional [catalog matching tools](https://docs.astropy.org/en/stable/coordinates/matchsep.html#matching-catalogs) are available from the
Astropy coordinates package.

Fortunately, people have worked on this problem, and the Gaia
database includes cross-matching tables that suggest a best neighbor
in the Pan-STARRS catalog for many stars in the Gaia catalog.

[This document describes the cross matching process](https://gea.esac.esa.int/archive/documentation/GDR2/Catalogue_consolidation/chap_cu9val_cu9val/ssec_cu9xma/sssec_cu9xma_extcat.html).
Briefly, it uses a cone search to find possible matches in
approximately the right position, then uses attributes like color and
magnitude to choose pairs of observations most likely to be the same
star.


## The best neighbor table

So the hard part of cross-matching has been done for us.  Using the
results is a little tricky, but it gives us a chance to learn about
one of the most important tools for working with databases: "joining"
tables.

A "join" is an operation where you match up records from
one table with records from another table using as a "key" a piece of
information that is common to both tables, usually some kind of ID
code.

In this example:

- Stars in the Gaia dataset are identified by `source_id`.

- Stars in the Pan-STARRS dataset are identified by `obj_id`.

For each candidate star we have selected so far, we have the
`source_id`; the goal is to find the `obj_id` for the same star
in the Pan-STARRS catalog.

To do that we will:

1. Use the `JOIN` operator to look up each Pan-STARRS `obj_id` for the stars
  we are interested in in the` panstarrs1_best_neighbour` table using the `source_id`s
  that we have already identified.

2. Use the `JOIN` operator again to look up the Pan-STARRS photometry for these stars
  in the `panstarrs1_original_valid` table using the` obj_ids` we just identified.

Before we get to the `JOIN` operation, we will explore these tables.

## 🔔 British vs American Spelling of Neighbour

The Gaia database was created and is maintained by the European Space Astronomy Center.
For this reason, the table spellings use the British spelling of neighbour
(with a "u"). Do not forget to include it in your table names in the queries below.


Here is the metadata for `panstarrs1_best_neighbour`.

In [3]:
ps_best_neighbour_meta = Gaia.load_table('gaiadr2.panstarrs1_best_neighbour')
print(ps_best_neighbour_meta)

Retrieving table 'gaiadr2.panstarrs1_best_neighbour'


And here are the columns.

In [6]:
for column in ps_best_neighbour_meta.columns:
    print(column.name)

source_id
original_ext_source_id
angular_distance
number_of_neighbours
number_of_mates
best_neighbour_multiplicity
gaia_astrometric_params


Here is the [documentation for these
variables](https://gea.esac.esa.int/archive/documentation/GDR2/Gaia_archive/chap_datamodel/sec_dm_crossmatches/ssec_dm_panstarrs1_best_neighbour.html).

The ones we will use are:

- `source_id`, which we will match up with `source_id` in the Gaia table.

- `best_neighbour_multiplicity`, which indicates how many sources in
  Pan-STARRS are matched with the same probability to this source in Gaia.

- `number_of_mates`, which indicates the number of *other* sources in
  Gaia that are matched with the same source in Pan-STARRS.

- `original_ext_source_id`, which we will match up with `obj_id` in
  the Pan-STARRS table.

Ideally, `best_neighbour_multiplicity` should be 1 and `number_of_mates`
should be 0; in that case, there is a one-to-one match between the
source in Gaia and the corresponding source in Pan-STARRS.

## 🔔 Number of neighbors

The table also contains `number_of_neighbours` which is the
number of stars in Pan-STARRS that match in terms of position, before
using other criteria to choose the most likely match.  But we are more
interested in the final match, using both criteria.

Here is a query that selects these columns and returns the first 5 rows.

In [8]:
ps_best_neighbour_query = """SELECT 
TOP 5
source_id, best_neighbour_multiplicity, number_of_mates, original_ext_source_id
FROM gaiadr2.panstarrs1_best_neighbour"""

In [9]:
ps_best_neighbour_job = Gaia.launch_job_async(ps_best_neighbour_query)

INFO: Query finished. [astroquery.utils.tap.core]


In [10]:
ps_best_neighbour_results = ps_best_neighbour_job.get_results()
ps_best_neighbour_results

<Table length=5>
     SOURCE_ID      best_neighbour_multiplicity number_of_mates original_ext_source_id
       int64                   int16                 int16              int64         
------------------- --------------------------- --------------- ----------------------
6745938972433480704                           1               0      69742925668851205
6030466788955954048                           1               0      69742509325691172
6756488099308169600                           1               0      69742879438541228
6700154994715046016                           1               0      69743055581721207
6757061941303252736                           1               0      69742856540241198

## The Pan-STARRS table

Now that we know the Pan-STARRS `obj_id`, we are ready to match this to the photometry in the
`panstarrs1_original_valid` table. Here is the metadata for the table that contains the Pan-STARRS data.

In [11]:
ps_valid_meta = Gaia.load_table('gaiadr2.panstarrs1_original_valid')

Retrieving table 'gaiadr2.panstarrs1_original_valid'


In [12]:
print(ps_valid_meta)

TAP Table name: gaiadr2.gaiadr2.panstarrs1_original_valid
Description: The Panoramic Survey Telescope and Rapid Response System (Pan-STARRS) is
a system for wide-field astronomical imaging developed and operated by
the Institute for Astronomy at the University of Hawaii. Pan-STARRS1
(PS1) is the first part of Pan-STARRS to be completed and is the basis
for Data Release 1 (DR1). The PS1 survey used a 1.8 meter telescope and
its 1.4 Gigapixel camera to image the sky in five broadband filters (g,
r, i, z, y).

The current table contains a filtered subsample of the 10 723 304 629
entries listed in the original ObjectThin table.
We used only ObjectThin and MeanObject tables to extract
panstarrs1OriginalValid table, this means that objects detected only in
stack images are not included here. The main reason for us to avoid the
use of objects detected in stack images is that their astrometry is not
as good as the mean objects astrometry: “The stack positions (raStack,
decStack) have considera

And here are the columns.

In [13]:
for column in ps_valid_meta.columns:
    print(column.name)

obj_name
obj_id
ra
dec
ra_error
dec_error
epoch_mean
g_mean_psf_mag
g_mean_psf_mag_error
g_flags
r_mean_psf_mag
r_mean_psf_mag_error
r_flags
i_mean_psf_mag
i_mean_psf_mag_error
i_flags
z_mean_psf_mag
z_mean_psf_mag_error
z_flags
y_mean_psf_mag
y_mean_psf_mag_error
y_flags
n_detections
zone_id
obj_info_flag
quality_flag


Here is the [documentation for these variables](https://gea.esac.esa.int/archive/documentation/GDR2/Gaia_archive/chap_datamodel/sec_dm_external_catalogues/ssec_dm_panstarrs1_original_valid.html) .

The ones we will use are:

- `obj_id`, which we will match up with `original_ext_source_id` in
  the best neighbor table.

- `g_mean_psf_mag`, which contains mean magnitude from the `g` filter.

- `i_mean_psf_mag`, which contains mean magnitude from the `i` filter.

Here is a query that selects these variables and returns the first 5 rows.

In [14]:
ps_valid_query = """SELECT 
TOP 5
obj_id, g_mean_psf_mag, i_mean_psf_mag 
FROM gaiadr2.panstarrs1_original_valid
"""

## Joining tables
---

The following figure shows how these tables are related.

- The orange circles and arrows represent the first `JOIN` operation,
  which takes each `source_id` in the Gaia table and finds the same
  value of `source_id` in the best neighbor table.

- The blue circles and arrows represent the second `JOIN` operation,
  which takes each `original_ext_source_id` in the best neighbor table and finds
  the same value of `obj_id` in the PanSTARRS photometry table.

There is no guarantee that the corresponding rows of these tables are
in the same order, so the `JOIN` operation involves some searching.
However, ADQL/SQL databases are implemented in a way that makes this
kind of search efficient.
If you are curious, you can [read more about
it](https://chartio.com/learn/databases/how-does-indexing-work/).

<div>
<img src="https://datacarpentry.org/astronomy-python/fig/join.png" width="1000"/>
</div>


Now we will get to the details of performing a `JOIN` operation.

We are about to build a complex query using software that doesn't provide us with any helpful information for debugging.
For this reason we are going to start with a simplified version of what we want to do until we are sure  we are joining
the tables correctly, then we will slowly add more layers of complexity, checking at each stage that our query still works.
As a starting place, we will go all the way back to the cone search from lesson 2.

In [15]:
test_cone_query = """SELECT 
TOP 10 
source_id
FROM gaiadr2.gaia_source
WHERE 1=CONTAINS(
  POINT(ra, dec),
  CIRCLE(88.8, 7.4, 0.08333333))
"""

And we will run it, to make sure we have a working query to build on.


In [16]:
test_cone_job = Gaia.launch_job(test_cone_query)

In [17]:
test_cone_results = test_cone_job.get_results()

In [18]:
test_cone_results

<Table length=10>
     SOURCE_ID     
       int64       
-------------------
3322773965056065536
3322773758899157120
3322774068134271104
3322773930696320512
3322774377374425728
3322773724537891456
3322773724537891328
3322773930696321792
3322773724537890944
3322773930696322176

Now we can start adding features.
First, we will replace `source_id` with the format specifier `columns` so that we can alter what columns we
want to return without having to modify our base query:

In [19]:
cone_base_query = """SELECT 
{columns}
FROM gaiadr2.gaia_source
WHERE 1=CONTAINS(
  POINT(ra, dec),
  CIRCLE(88.8, 7.4, 0.08333333))
"""

As a reminder, here are the columns we want from the Gaia table:

```python
columns = 'source_id, ra, dec, pmra, pmdec'
```


In [20]:
columns = 'source_id, ra, dec, pmra, pmdec'

cone_query = cone_base_query.format(columns=columns)
print(cone_query)

SELECT 
source_id, ra, dec, pmra, pmdec
FROM gaiadr2.gaia_source
WHERE 1=CONTAINS(
  POINT(ra, dec),
  CIRCLE(88.8, 7.4, 0.08333333))



We run the query again.

In [21]:
cone_job = Gaia.launch_job_async(cone_query)

INFO: Query finished. [astroquery.utils.tap.core]


In [22]:
cone_results = cone_job.get_results()
cone_results

<Table length=594>
     SOURCE_ID              ra               dec                 pmra                pmdec       
                           deg               deg               mas / yr             mas / yr     
       int64             float64           float64             float64              float64      
------------------- ----------------- ------------------ -------------------- -------------------
3322773965056065536 88.78178020183375  7.334936530583141   0.2980633722108194 -2.5057036964736907
3322773758899157120 88.83227057144585  7.325577341429926                   --                  --
3322774068134271104  88.8206092188033  7.353158142762173  -1.1065462654445488 -1.5260889445858044
3322773930696320512 88.80843339290348  7.334853162299928   2.6074384482375215 -0.9292104395445717
3322774377374425728 88.86806108182265  7.371287731275939   3.9555477866915383 -3.8676624830902435
3322773724537891456 88.81308602813434   7.32488574492059    51.34995462741039 -33.078133430952086
3322773724537891328 88.81570329208743 7.3223019772324855   1.9389988498951845  0.3110526931576576
3322773930696321792  88.8050736770331  7.332371472206583    2.264014834476311  1.0772755505138008
3322773724537890944 88.81241651540533  7.327864052479726 -0.36003627434304625  -6.393939291541333
3322773930696322176 88.80128682574824  7.334292036448643                   --                  --
3322774171213468288 88.85926451433244  7.347883547366655                   --                  --
3322774549172100864 88.83436553681659  7.381718098527705  -2.8346862156210197 -10.041670801563173
3322774480453643776 88.84560455437382   7.37235454153855  -2.3607769454630194  0.3119094236284546
3322774549173120768 88.83656567295009  7.379145614749243   1.1879010806834702  0.8015608679926151
3322774102494018816 88.80206135956955  7.352773486341117 -0.44875204757926507 0.06388813412223136
3322774480453644544 88.84518455090614  7.369528614293598   1.0487255252218055 -3.2216341675741775
3322774102495010432 88.79716654349421  7.352131803476284                   --                  --
3322773896336588288 88.79204096615393   7.32650520319297   -2.032795831142071   3.667053981953014
3322774514810855168 88.82561554658879  7.370396362874608    2.050310050020124  -2.049474519300541
3322774377372872192 88.87582672533853  7.366118779323389   1.8127015837108136  0.7001194505950585
3322774583531302528 88.85999672738541  7.382293774944985  -0.6623070684058512 -2.3442341897855243
3322774652250779392  88.8524947323721  7.388899986262633   1.6972219448237815  1.6485794601186199
3322774686610517248 88.85472110026933  7.390063397753613                   --                  --
3322774068135266176  88.8235301157897  7.347705160722182  -3.7433564817590645 0.31886651205178934
3322773857683389824 88.83996292432077 7.3445665848082236   2.6534378827450373   -8.01774392396456
3322774583531304832 88.85867590829105  7.376481786908238   2.6366028834302653  1.8134315265562975
3322773827616107136 88.83071130581675  7.335176978367904   1.4782268393421794  -4.210984137177093
                ...               ...                ...                  ...                 ...
3322962325141750016 88.83569860807606  7.400019185834357    2.232528951329276 0.06553486501513013
3322963424653377664 88.79010085612454  7.440780293746405  -0.5665536634320726   1.131514742770427
3322962325141749760 88.83734552112712  7.398926808815748  0.34393181491939895  0.5183421380693837
3322962977978513408  88.7442802483588  7.412942724151012   -6.554444262076663   7.063793489936718
3322962118985054336 88.75955522181778  7.395570338211606    3.725957752362235 -2.2398692278343715
3322963184136943872  88.7309524802499   7.42480485556539    2.305446276420357 -11.326745896862978
3322963768250764288 88.75839741640878   7.45788588719871 -0.14370138848752928  0.3586740564258863
3322961844105448192  88.7811364509495  7.363879864829356   1.0613332159564797 0.18795846530326532
3322963768250764416 88.75896332337557  7.456741652211021   -1

## Adding the best neighbor table

Now we are ready for the first join.
The join operation requires two clauses:

- `JOIN` specifies the name of the table we want to join with, and

- `ON` specifies how we will match up rows between the tables.

In this example, we join with `gaiadr2.panstarrs1_best_neighbour AS best`, which means we can refer to the best neighbor table with the
abbreviated name `best`, which will save us a lot of typing. Similarly,
we will be referring to the `gaiadr2.gaia_source` table by the abbreviated name `gaia`.

The `ON` clause indicates that we will match up the `source_id`
column from the Gaia table with the `source_id` column from the best
neighbor table.

```python
neighbours_base_query = """SELECT 
{columns}
FROM gaiadr2.gaia_source AS gaia
JOIN gaiadr2.panstarrs1_best_neighbour AS best
  ON gaia.source_id = best.source_id
WHERE 1=CONTAINS(
  POINT(gaia.ra, gaia.dec),
  CIRCLE(88.8, 7.4, 0.08333333))
"""
```

In [24]:
neighbours_base_query = """SELECT 
{columns}
FROM gaiadr2.gaia_source AS gaia
JOIN gaiadr2.panstarrs1_best_neighbour AS best
  ON gaia.source_id = best.source_id
WHERE 1=CONTAINS(
  POINT(gaia.ra, gaia.dec),
  CIRCLE(88.8, 7.4, 0.08333333))
"""

## 🔔 SQL detail

In this example, the `ON` column has the same name in both tables, so we could replace the `ON` clause
with a simpler [`USING`clause](https://docs.oracle.com/javadb/10.8.3.0/ref/rrefsqljusing.html):

```sql
USING(source_id)
```


Now that there is more than one table involved, we can't use simple
column names any more; we have to use **qualified column names**.
In other words, we have to specify which table each column is in.
The column names do not have to be the same and, in fact, in the next join they will not be.
That is one of the reasons that we explicitly specify them.
Here is the complete query, including the columns we want from the Gaia
and best neighbor tables. Here you can start to see that using the abbreviated names
is making our query easier to read and requires less typing for us. In addition to the
spatial coordinates and proper motion, we are going to return the `best_neighbour_multiplicity`
and `number_of_mates` columns from the `panstarrs1_best_neighbour` table in order to evaluate the quality
of the data that we are using by evaluating the number of one-to-one matches between the catalogs.
Recall that `best_neighbour_multiplicity` tells us the number of PanSTARRs
objects that match a Gaia object and `number_of_mates` tells us the number of Gaia objects that match a
PanSTARRs object.


In [25]:
column_list_neighbours = ['gaia.source_id',
               'gaia.ra',
               'gaia.dec',
               'gaia.pmra',
               'gaia.pmdec',
               'best.best_neighbour_multiplicity',
               'best.number_of_mates',
              ]
columns = ', '.join(column_list_neighbours)

neighbours_query = neighbours_base_query.format(columns=columns)
print(neighbours_query)

SELECT 
gaia.source_id, gaia.ra, gaia.dec, gaia.pmra, gaia.pmdec, best.best_neighbour_multiplicity, best.number_of_mates
FROM gaiadr2.gaia_source AS gaia
JOIN gaiadr2.panstarrs1_best_neighbour AS best
  ON gaia.source_id = best.source_id
WHERE 1=CONTAINS(
  POINT(gaia.ra, gaia.dec),
  CIRCLE(88.8, 7.4, 0.08333333))



In [26]:
neighbours_job = Gaia.launch_job_async(neighbours_query)

INFO: Query finished. [astroquery.utils.tap.core]


In [27]:
neighbours_results = neighbours_job.get_results()
neighbours_results

<Table length=490>
     SOURCE_ID              ra               dec                 pmra                pmdec        best_neighbour_multiplicity number_of_mates
                           deg               deg               mas / yr             mas / yr                                                 
       int64             float64           float64             float64              float64                  int16                 int16     
------------------- ----------------- ------------------ -------------------- ------------------- --------------------------- ---------------
3322773965056065536 88.78178020183375  7.334936530583141   0.2980633722108194 -2.5057036964736907                           1               0
3322774068134271104  88.8206092188033  7.353158142762173  -1.1065462654445488 -1.5260889445858044                           1               0
3322773930696320512 88.80843339290348  7.334853162299928   2.6074384482375215 -0.9292104395445717                           1               0
3322774377374425728 88.86806108182265  7.371287731275939   3.9555477866915383 -3.8676624830902435                           1               0
3322773724537891456 88.81308602813434   7.32488574492059    51.34995462741039 -33.078133430952086                           1               0
3322773724537891328 88.81570329208743 7.3223019772324855   1.9389988498951845  0.3110526931576576                           1               0
3322773930696321792  88.8050736770331  7.332371472206583    2.264014834476311  1.0772755505138008                           1               0
3322773724537890944 88.81241651540533  7.327864052479726 -0.36003627434304625  -6.393939291541333                           1               0
3322773930696322176 88.80128682574824  7.334292036448643                   --                  --                           1               0
3322774549172100864 88.83436553681659  7.381718098527705  -2.8346862156210197 -10.041670801563173                           1               0
3322774480453643776 88.84560455437382   7.37235454153855  -2.3607769454630194  0.3119094236284546                           1               0
3322774549173120768 88.83656567295009  7.379145614749243   1.1879010806834702  0.8015608679926151                           1               0
3322774102494018816 88.80206135956955  7.352773486341117 -0.44875204757926507 0.06388813412223136                           1               0
3322774480453644544 88.84518455090614  7.369528614293598   1.0487255252218055 -3.2216341675741775                           1               0
3322773896336588288 88.79204096615393   7.32650520319297   -2.032795831142071   3.667053981953014                           1               0
3322774514810855168 88.82561554658879  7.370396362874608    2.050310050020124  -2.049474519300541                           1               0
3322774377372872192 88.87582672533853  7.366118779323389   1.8127015837108136  0.7001194505950585                           1               0
3322774583531302528 88.85999672738541  7.382293774944985  -0.6623070684058512 -2.3442341897855243                           1               0
3322774652250779392  88.8524947323721  7.388899986262633   1.6972219448237815  1.6485794601186199                           1               0
3322774686610517248 88.85472110026933  7.390063397753613                   --                  --                           1               0
3322774068135266176  88.8235301157897  7.347705160722182  -3.7433564817590645 0.31886651205178934                           1               0
3322773857683389824 88.83996292432077 7.3445665848082236   2.6534378827450373   -8.01774392396456                           1               0
3322774583531304832 88.85867590829105  7.376481786908238   2.6366028834302653  1.8134315265562975                           1               0
3322773827616107136 88.83071130581675  7.335176978367904   1.4782268393421794  -4.210984137177093                           1               0
33227

This result has fewer rows than the previous result.
That is because there are sources in the Gaia table with no
corresponding source in the Pan-STARRS table.

By default, the result of the join only includes rows where the same
`source_id` appears in both tables.
This default is called an "inner" join because the results include
only the intersection of the two tables.
[You can read about the other kinds of join
here](https://www.geeksforgeeks.org/sql-join-set-1-inner-left-right-and-full-joins/).

## <p style="background-color: #f5df18; padding: 10px;"> 🛑 Adding the Pan-STARRS table (15 Minutes) </p>
---

Now we are ready to bring in the Pan-STARRS table.  Starting with the
previous query, add a second `JOIN` clause that joins with
`gaiadr2.panstarrs1_original_valid`, gives it the abbreviated name
`ps`, and matches `original_ext_source_id` from the best neighbor
table with `obj_id` from the Pan-STARRS table.

Add `g_mean_psf_mag` and `i_mean_psf_mag` to the column list, and run the query.
The result should contain 490 rows and 9 columns.

In [28]:
join_solution_query_base = """SELECT 
{columns}
FROM gaiadr2.gaia_source as gaia
JOIN gaiadr2.panstarrs1_best_neighbour as best
  ON gaia.source_id = best.source_id
JOIN gaiadr2.panstarrs1_original_valid as ps
  ON best.original_ext_source_id = ps.obj_id
WHERE 1=CONTAINS(
  POINT(gaia.ra, gaia.dec),
  CIRCLE(88.8, 7.4, 0.08333333))
"""

column_list = ['gaia.source_id',
               'gaia.ra',
               'gaia.dec',
               'gaia.pmra',
               'gaia.pmdec',
               'best.best_neighbour_multiplicity',
               'best.number_of_mates',
               'ps.g_mean_psf_mag',
               'ps.i_mean_psf_mag']

columns = ', '.join(column_list)

join_solution_query = join_solution_query_base.format(columns=columns)
print(join_solution_query)

join_solution_job = Gaia.launch_job_async(join_solution_query)
join_solution_results = join_solution_job.get_results()
join_solution_results

SELECT 
gaia.source_id, gaia.ra, gaia.dec, gaia.pmra, gaia.pmdec, best.best_neighbour_multiplicity, best.number_of_mates, ps.g_mean_psf_mag, ps.i_mean_psf_mag
FROM gaiadr2.gaia_source as gaia
JOIN gaiadr2.panstarrs1_best_neighbour as best
  ON gaia.source_id = best.source_id
JOIN gaiadr2.panstarrs1_original_valid as ps
  ON best.original_ext_source_id = ps.obj_id
WHERE 1=CONTAINS(
  POINT(gaia.ra, gaia.dec),
  CIRCLE(88.8, 7.4, 0.08333333))

INFO: Query finished. [astroquery.utils.tap.core]


<Table length=490>
     SOURCE_ID              ra               dec                 pmra                pmdec        best_neighbour_multiplicity number_of_mates  g_mean_psf_mag   i_mean_psf_mag 
                           deg               deg               mas / yr             mas / yr                                                                         mag       
       int64             float64           float64             float64              float64                  int16                 int16          float64          float64     
------------------- ----------------- ------------------ -------------------- ------------------- --------------------------- --------------- ---------------- ----------------
3322773965056065536 88.78178020183375  7.334936530583141   0.2980633722108194 -2.5057036964736907                           1               0 19.9431991577148 17.4221992492676
3322774068134271104  88.8206092188033  7.353158142762173  -1.1065462654445488 -1.5260889445858044                           1               0 18.6212005615234 16.6007995605469
3322773930696320512 88.80843339290348  7.334853162299928   2.6074384482375215 -0.9292104395445717                           1               0               -- 20.2203998565674
3322774377374425728 88.86806108182265  7.371287731275939   3.9555477866915383 -3.8676624830902435                           1               0 18.0676002502441 16.9762001037598
3322773724537891456 88.81308602813434   7.32488574492059    51.34995462741039 -33.078133430952086                           1               0 20.1907005310059 17.8700008392334
3322773724537891328 88.81570329208743 7.3223019772324855   1.9389988498951845  0.3110526931576576                           1               0 22.6308002471924 19.6004009246826
3322773930696321792  88.8050736770331  7.332371472206583    2.264014834476311  1.0772755505138008                           1               0 21.2119998931885 18.3528003692627
3322773724537890944 88.81241651540533  7.327864052479726 -0.36003627434304625  -6.393939291541333                           1               0 20.8094005584717 18.1343002319336
3322773930696322176 88.80128682574824  7.334292036448643                   --                  --                           1               0 19.7306003570557               --
3322774549172100864 88.83436553681659  7.381718098527705  -2.8346862156210197 -10.041670801563173                           1               0               -- 16.1383991241455
3322774480453643776 88.84560455437382   7.37235454153855  -2.3607769454630194  0.3119094236284546                           1               0 16.8370990753174 15.8711004257202
3322774549173120768 88.83656567295009  7.379145614749243   1.1879010806834702  0.8015608679926151                           1               0 17.8430004119873 16.4561004638672
3322774102494018816 88.80206135956955  7.352773486341117 -0.44875204757926507 0.06388813412223136                           1               0 18.5690002441406 16.2737998962402
3322774480453644544 88.84518455090614  7.369528614293598   1.0487255252218055 -3.2216341675741775                           1               0 16.4340000152588 15.6140003204346
3322773896336588288 88.79204096615393   7.32650520319297   -2.032795831142071   3.667053981953014                           1               0               -- 17.6795997619629
3322774514810855168 88.82561554658879  7.370396362874608    2.050310050020124  -2.049474519300541                           1               0 18.0564994812012 16.6308002471924
3322774377372872192 88.87582672533853  7.366118779323389   1.8127015837108136  0.7001194505950585                           1               0               -- 17.7392997741699
3322774583531302528 88.85999672738541  7.382293774944985  -0.6623070684058512 -2.3442341897855243                           1               0 19.4053001403809 17.8836994171143
3322774652250779392  88.8524947323721  7.388899986262633   1.6972219448237815  1.6485794601186199            

## Selecting by coordinates and proper motion

We are now going to replace the cone search with the GD-1 selection that we built in previous episodes.
We will start by making sure that our previous query works, then add in the `JOIN`.
Now we will bring in the `WHERE` clause from the previous episode, which
selects sources based on parallax, BP-RP color, sky coordinates, and
proper motion.

Here is `candidate_coord_pm_query_base` from the previous episode.

```python
candidate_coord_pm_query_base = """SELECT 
{columns}
FROM gaiadr2.gaia_source
WHERE parallax < 1
  AND bp_rp BETWEEN -0.75 AND 2 
  AND 1 = CONTAINS(POINT(ra, dec), 
                   POLYGON({sky_point_list}))
  AND pmra BETWEEN {pmra_min} AND  {pmra_max}
  AND pmdec BETWEEN {pmdec_min} AND {pmdec_max}
"""
```

Now we can assemble the query using the sky point list and proper motion range we compiled in episode 5.

In [29]:
candidate_coord_pm_query_base = """SELECT 
{columns}
FROM gaiadr2.gaia_source
WHERE parallax < 1
  AND bp_rp BETWEEN -0.75 AND 2 
  AND 1 = CONTAINS(POINT(ra, dec), 
                   POLYGON({sky_point_list}))
  AND pmra BETWEEN {pmra_min} AND  {pmra_max}
  AND pmdec BETWEEN {pmdec_min} AND {pmdec_max}
"""

In [30]:
columns = 'source_id, ra, dec, pmra, pmdec'

candidate_coord_pm_query = candidate_coord_pm_query_base.format(columns=columns,
                            sky_point_list=sky_point_list,
                            pmra_min=pmra_min,
                            pmra_max=pmra_max,
                            pmdec_min=pmdec_min,
                            pmdec_max=pmdec_max)

print(candidate_coord_pm_query)

SELECT 
source_id, ra, dec, pmra, pmdec
FROM gaiadr2.gaia_source
WHERE parallax < 1
  AND bp_rp BETWEEN -0.75 AND 2 
  AND 1 = CONTAINS(POINT(ra, dec), 
                   POLYGON(135.306, 8.39862, 126.51, 13.4449, 163.017, 54.2424, 172.933, 46.4726, 135.306, 8.39862))
  AND pmra BETWEEN -6.7 AND  -3
  AND pmdec BETWEEN -14.31 AND -11.2



We run it to make sure we are starting with a working query.

In [31]:
candidate_coord_pm_job = Gaia.launch_job_async(candidate_coord_pm_query)

INFO: Query finished. [astroquery.utils.tap.core]


In [32]:
candidate_coord_pm_results = candidate_coord_pm_job.get_results()
candidate_coord_pm_results

<Table length=8409>
    SOURCE_ID              ra                dec                 pmra               pmdec       
                          deg                deg               mas / yr            mas / yr     
      int64             float64            float64             float64             float64      
------------------ ------------------ ------------------ ------------------- -------------------
635559124339440000 137.58671691646745   19.1965441084838  -3.770521900009566 -12.490481778113859
635860218726658176  138.5187065217173  19.09233926905897  -5.941679495793577 -11.346409129876392
635674126383965568  138.8428741026386 19.031798198627634 -3.8970011609340207 -12.702779525389634
635614168640132864 139.59219748145836 18.807955539071433  -3.309602916796381 -13.708904908478631
635523531945742592  137.1439946395468 18.845588306639165 -3.9036930274876585 -11.634380634589444
635821843194387840 139.88094034815086  19.62185456718988  -6.544201177153814  -12.55978220563274
635821980633338624  139.9080311966138 19.638380144036358  -4.045391546526935 -11.966804519012523
635551706931167104 138.04665586038192 19.248909662830798  -6.224595114220405 -12.224246333795001
635518889086133376  137.2374229207837   18.7428630711791 -3.3186800714801046 -12.710314902969365
635694467349085056 139.06182692172476  19.06744237567232 -3.0674889862265795   -11.4175270465402
635580294233854464 139.44127324251306 18.465885053778493   -4.45999663908048 -12.803774239857741
635616539462230400   139.748253143578 18.816930070010645  -3.764502655934397 -12.666704724957508
635500953302563712 137.47969724876137 18.589457707583403  -3.194660711570627 -12.948727714046099
635602039653056000 139.18302300031363 18.676799274156195 -3.2671051023818496  -11.67647140219096
635598607974369792 139.20920023089508 18.624132868942702  -6.124445176881091 -12.833824027100611
635641592007083520  138.9435110321861 18.610237707999868  -3.791613660715501  -12.86516451722882
635684266802035712 139.17274439579512   19.1082761558709  -4.407971465526585 -12.764441856174665
635517274178016640 137.62830801613939 18.948650358846358  -6.458846314736528 -14.097877513517432
635567989151968768 137.75105460698236 19.360622924269965  -4.049188459972398 -13.042919254110684
635684713478631168  139.2153044491863 19.123124494930934  -4.673761807938962 -13.267407044908088
635850945892748672 139.86542888472115 20.011312663154804  -3.786655365804428  -14.28415600718206
635600532119713664 139.22869949616816 18.685939084485494 -3.9742788217925122 -12.342426623384245
635765596302772736 140.04734758156198 19.406283833620364  -3.763802631170119 -11.516039849448802
635620701286087424  139.6862822364899 18.881113520820737 -3.3558418379356474 -12.925224236202752
635740204456128896 139.98945182326287  19.27694306975315 -3.5117247908083487 -11.701695774207732
635745388481042176  139.8715426069336  19.36739472431735  -6.340716495553142 -12.869365081573445
635832357274302208  140.1301532714717 19.821917114066117  -4.055981152196889 -13.024120128840579
               ...                ...                ...                 ...                 ...
612155542602200320 135.20109084925417 18.103606015244775  -6.422352169700357 -12.928463131169904
611955946881374592 135.14820752449802 17.426270937913014 -6.0227363644651914 -11.459642240642477
612177975215097344  135.6728203544916 18.358672432072115 -3.5237019478067504 -13.537073125325234
612107606473231232 136.03756654138073 18.041440826429486  -5.549022301194379 -12.160712951323148
612037924926682112 134.84445715353883  17.72323933639592  -4.269227658500999  -11.49981064531317
612270407207552640 134.70922405308724  18.29130802076034  -4.839939127439933 -11.725099676826757
612396060770518272 135.66282067939025 18.872963154416713  -3.455761609183599 -12.332495145103103
612352320823604736 135.05170970268458 18.367568252433077 -3.2660135577877325 -13.132871599475262
612298513473304832 133.82579580380593 18.159803159638237  -5.861963949768889 -12.495341645865798
612

## <p style="background-color: #f5df18; padding: 10px;"> 🛑 Exercise (15 Minutes) </p>
---

Create a new query base called `candidate_join_query_base` that combines the `WHERE`
clauses from the previous query with the `JOIN` clauses for the best
neighbor and Pan-STARRS tables.
Format the query base using the column names in `column_list`, and
call the result `candidate_join_query`.

Hint: Make sure you use qualified column names everywhere!

Run your query and download the results.  The table you get should
have 4300 rows and 9 columns.

In [33]:
candidate_join_query_base = """
SELECT 
{columns}
FROM gaiadr2.gaia_source as gaia
JOIN gaiadr2.panstarrs1_best_neighbour as best
  ON gaia.source_id = best.source_id
JOIN gaiadr2.panstarrs1_original_valid as ps
  ON best.original_ext_source_id = ps.obj_id
WHERE parallax < 1
  AND bp_rp BETWEEN -0.75 AND 2 
  AND 1 = CONTAINS(POINT(gaia.ra, gaia.dec), 
                   POLYGON({sky_point_list}))
  AND gaia.pmra BETWEEN {pmra_min} AND  {pmra_max}
  AND gaia.pmdec BETWEEN {pmdec_min} AND {pmdec_max}
"""

columns = ', '.join(column_list)

candidate_join_query = candidate_join_query_base.format(columns=columns,
                            sky_point_list= sky_point_list,
                            pmra_min=pmra_min,
                            pmra_max=pmra_max,
                            pmdec_min=pmdec_min,
                            pmdec_max=pmdec_max)
print(candidate_join_query)


candidate_join_job = Gaia.launch_job_async(candidate_join_query)
candidate_table = candidate_join_job.get_results()
candidate_table


SELECT 
gaia.source_id, gaia.ra, gaia.dec, gaia.pmra, gaia.pmdec, best.best_neighbour_multiplicity, best.number_of_mates, ps.g_mean_psf_mag, ps.i_mean_psf_mag
FROM gaiadr2.gaia_source as gaia
JOIN gaiadr2.panstarrs1_best_neighbour as best
  ON gaia.source_id = best.source_id
JOIN gaiadr2.panstarrs1_original_valid as ps
  ON best.original_ext_source_id = ps.obj_id
WHERE parallax < 1
  AND bp_rp BETWEEN -0.75 AND 2 
  AND 1 = CONTAINS(POINT(gaia.ra, gaia.dec), 
                   POLYGON(135.306, 8.39862, 126.51, 13.4449, 163.017, 54.2424, 172.933, 46.4726, 135.306, 8.39862))
  AND gaia.pmra BETWEEN -6.7 AND  -3
  AND gaia.pmdec BETWEEN -14.31 AND -11.2

INFO: Query finished. [astroquery.utils.tap.core]


<Table length=4300>
    SOURCE_ID              ra                dec                 pmra               pmdec        best_neighbour_multiplicity number_of_mates  g_mean_psf_mag   i_mean_psf_mag 
                          deg                deg               mas / yr            mas / yr                                                                         mag       
      int64             float64            float64             float64             float64                  int16                 int16          float64          float64     
------------------ ------------------ ------------------ ------------------- ------------------- --------------------------- --------------- ---------------- ----------------
635860218726658176  138.5187065217173  19.09233926905897  -5.941679495793577 -11.346409129876392                           1               0 17.8978004455566 17.5174007415771
635674126383965568  138.8428741026386 19.031798198627634 -3.8970011609340207 -12.702779525389634                           1               0 19.2873001098633 17.6781005859375
635614168640132864 139.59219748145836 18.807955539071433  -3.309602916796381 -13.708904908478631                           1               0 16.1515998840332 14.6662998199463
635694467349085056 139.06182692172476  19.06744237567232 -3.0674889862265795   -11.4175270465402                           1               0 18.6709003448486 17.9841995239258
635598607974369792 139.20920023089508 18.624132868942702  -6.124445176881091 -12.833824027100611                           1               0 16.5223999023438 16.1375007629395
635850945892748672 139.86542888472115 20.011312663154804  -3.786655365804428  -14.28415600718206                           1               0 16.5174999237061 16.0450000762939
635600532119713664 139.22869949616816 18.685939084485494 -3.9742788217925122 -12.342426623384245                           1               0 20.4505996704102 19.5177001953125
635620701286087424  139.6862822364899 18.881113520820737 -3.3558418379356474 -12.925224236202752                           1               0 20.4113006591797 19.5032997131348
635740204456128896 139.98945182326287  19.27694306975315 -3.5117247908083487 -11.701695774207732                           1               0 15.8247995376587 15.2495002746582
635832357274302208  140.1301532714717 19.821917114066117  -4.055981152196889 -13.024120128840579                           1               0 20.4759006500244 18.8048000335693
635593552797727488 139.00070003257937  18.55590290002122  -4.306131802878042  -12.54195022893678                           1               0 20.6061992645264 18.8955001831055
635667426234947072 138.89750064175942 18.872682695055357  -5.157042432354842  -13.24136226011474                           1               0 19.8118000030518 17.9141998291016
635518889086125824 137.23270610620415  18.74116628926074  -4.465384592407755  -13.20837869612681                           1               0 22.1968994140625  19.910400390625
635672030439913088 138.75338841789633 18.969538186235194  -4.310354022480536 -11.845731163540334                           1               0 19.6383991241455 18.4435005187988
635830463193424256   139.999132312995 19.762252802658715  -6.584792205166673 -13.091381925525644                           1               0 19.7700996398926  18.096399307251
635785284432814976 139.59848476562738 19.319710275491186  -6.519450594788479 -11.639426669946435                           1               0 16.1424999237061 15.5797004699707
635724566479972736 140.03325344702228 19.111862759270924  -4.767166134623367 -12.434745246609625                           1               0 20.1289005279541 19.3251991271973
684873285997511296 134.88835179418987 20.655998230170262 -3.4510057757046795 -11.202379479112448                           1               0 21.3948993682861 19.3689994812012
685024434488326272 135.26750694153745  21.32978771858751  -3.493039346292803 -11.886709712504283                           1      

# Checking the match
---

To get more information about the matching process, we can inspect
`best_neighbour_multiplicity`, which indicates for each star in Gaia
how many stars in Pan-STARRS are equally likely matches.

In [34]:
candidate_table['best_neighbour_multiplicity']

<MaskedColumn name='best_neighbour_multiplicity' dtype='int16' description='Number of neighbours with same probability as best neighbour' length=4300>
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
...
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1
  1

Most of the values are `1`, which is good; that means
that for each candidate star we have identified exactly one source in
Pan-STARRS that is likely to be the same star.

To check whether there are any values other than `1`, we can convert
this column to a Pandas `Series` and use `describe`, which we saw in
in episode 3.

```python
multiplicity = pd.Series(candidate_table['best_neighbour_multiplicity'])
multiplicity.describe()
```

```output
count    4300.0
mean        1.0
std         0.0
min         1.0
25%         1.0
50%         1.0
75%         1.0
max         1.0
dtype: float64
```

In fact, `1` is the only value in the `Series`, so every candidate
star has a single best match.

## 🔔 Numpy Mask Warning

You may see a warning that ends with the following phrase:

`site-packages/numpy/lib/function_base.py:4650:`

`UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedColumn.`
`arr.partition(`

This is because astroquery is returning a table with masked columns (which are really fancy masked numpy arrays).
When we turn this column into a pandas Series, it maintains its mask. Describe calls numpy functions to perform statistics.
Numpy recently implemented this warning to let you know that the mask is not being considered in the calculation
its performing.


Similarly, `number_of_mates` indicates the number of *other* stars in
Gaia that match with the same star in Pan-STARRS.

In [35]:
mates = pd.Series(candidate_table['number_of_mates'])
mates.describe()

/Users/tb/anaconda3/lib/python3.11/site-packages/numpy/lib/function_base.py:4824: UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedColumn.
  arr.partition(


count    4300.0
mean        0.0
std         0.0
min         0.0
25%         0.0
50%         0.0
75%         0.0
max         0.0
dtype: float64

All values in this column are `0`, which means that for each match we
found in Pan-STARRS, there are no other stars in Gaia that also match.

# Saving the DataFrame
---

We can make a `DataFrame` from our Astropy `Table` and save our results so we can pick up where we left off
without running this query again. Once again, we will make use of our `make_dataframe` function.

In [36]:
candidate_df = make_dataframe(candidate_table)

The HDF5 file should already exist, so we'll add `candidate_df` to it.

In [37]:
filename = 'gd1_data.hdf'

candidate_df.to_hdf(filename, 'candidate_df')


/var/folders/b3/gxd2jb0s4bq1wknd7l1bnpg80000gn/T/ipykernel_15931/97204592.py:3: FutureWarning: Starting with pandas version 3.0 all arguments of to_hdf except for the argument 'path_or_buf' will be keyword-only.
  candidate_df.to_hdf(filename, 'candidate_df')


We can use `getsize` to confirm that the file exists and check the size:

In [38]:
from os.path import getsize

MB = 1024 * 1024
getsize(filename) / MB


15.428459167480469

# Another file format - CSV
--- 

Pandas can write a variety of other formats, [which you can read about
here](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html).
We won't cover all of them, but one other important one is
[CSV](https://en.wikipedia.org/wiki/Comma-separated_values), which
stands for "comma-separated values".

CSV is a plain-text format that can be read and written by pretty much
any tool that works with data.  In that sense, it is the "least common
denominator" of data formats.

However, it has an important limitation: some information about the
data gets lost in translation, notably the data types.  If you read a
CSV file from someone else, you might need some additional information
to make sure you are getting it right.

Also, CSV files tend to be big, and slow to read and write.

With those caveats, here is how to write one:

the kind of thing that can go wrong with CSV files.

In [39]:
candidate_df.to_csv('gd1_data.csv')

We can check the file size like this:

In [40]:
getsize('gd1_data.csv') / MB

0.8788290023803711

We can read the CSV file back like this:

In [41]:
read_back_csv = pd.read_csv('gd1_data.csv')

We will compare the first few rows of `candidate_df` and `read_back_csv`

In [42]:
candidate_df.head(3)

SOURCE_ID          ra        dec      pmra      pmdec  \
0  635860218726658176  138.518707  19.092339 -5.941679 -11.346409   
1  635674126383965568  138.842874  19.031798 -3.897001 -12.702780   
2  635614168640132864  139.592197  18.807956 -3.309603 -13.708905   

   best_neighbour_multiplicity  number_of_mates  g_mean_psf_mag  \
0                            1                0         17.8978   
1                            1                0         19.2873   
2                            1                0         16.1516   

   i_mean_psf_mag       phi1      phi2   pm_phi1   pm_phi2  
0       17.517401 -59.247330 -2.016078 -7.527126  1.748779  
1       17.678101 -59.133391 -2.306901 -7.560608 -0.741800  
2       14.666300 -58.938113 -3.024192 -8.062762 -1.869082

In [43]:
read_back_csv.head(3)

Unnamed: 0           SOURCE_ID          ra        dec      pmra      pmdec  \
0           0  635860218726658176  138.518707  19.092339 -5.941679 -11.346409   
1           1  635674126383965568  138.842874  19.031798 -3.897001 -12.702780   
2           2  635614168640132864  139.592197  18.807956 -3.309603 -13.708905   

   best_neighbour_multiplicity  number_of_mates  g_mean_psf_mag  \
0                            1                0         17.8978   
1                            1                0         19.2873   
2                            1                0         16.1516   

   i_mean_psf_mag       phi1      phi2   pm_phi1   pm_phi2  
0       17.517401 -59.247330 -2.016078 -7.527126  1.748779  
1       17.678101 -59.133391 -2.306901 -7.560608 -0.741800  
2       14.666300 -58.938113 -3.024192 -8.062762 -1.869082

# Summary

---

In this episode, we used database `JOIN` operations to select
photometry data for the stars we've identified as candidates to be in
GD-1.

In the next episode, we will use this data for a second round of
selection, identifying stars that have photometry data consistent with
GD-1.

# <p style="background-color: #f5df18; padding: 10px;"> 🗝️ Key points</p>

---

- Use `JOIN` operations to combine data from multiple tables in a database, using some kind of identifier to match up records from one table with records from another. This is another example of a practice we saw in the previous notebook, moving the computation to the data.
- For most applications, saving data in FITS or HDF5 is better than CSV.  FITS and HDF5 are binary formats, so the files are usually smaller, and they store metadata, so you don't lose anything when you read the file back.
- On the other hand, CSV is a 'least common denominator' format; that is, it can be read by practically any application that works with data.